x-pack/plugins/security/server/authentication/api_keys/api_keys.ts

```diff
@@ -329,4 +366,50 @@ export class APIKeys {
 
     throw new Error(`Unsupported scheme "${authorizationHeader.scheme}" for granting API Key`);
   }
+
+  private parseRoleDescriptorsWithKibanaPrivileges(createParams: CreateAPIKeyParams) {
+    let roleDescriptors = createParams.role_descriptors;
+
+    if (roleDescriptors) {
+      return roleDescriptors;
+    }
+
+    roleDescriptors = {};
+    const { kibana_role_descriptors: kibanaRoleDescriptors } = createParams;
+
+    const allValidationErrors: string[] = [];
+    if (kibanaRoleDescriptors) {
+      Object.entries(kibanaRoleDescriptors).forEach(([roleKey, roleDescriptor]) => {
+        const { validationErrors } = validateKibanaPrivileges(
+          this.kibanaFeatures,
+          roleDescriptor.kibana
+        );
+        allValidationErrors.push(...validationErrors);
+
+        const applications = transformPrivilegesToElasticsearchPrivileges(
+          this.applicationName,
+          roleDescriptor.kibana
+        );
+        if (applications.length > 0 && roleDescriptors) {
+          roleDescriptors[roleKey] = {
```

Comment:
`roleKey` is user controlled, so this could equal something like `__proto__`. I recommend either creating `roleDescriptors` with a null prototype, or changing the way we collect these descriptors